traits/observers/expressions.py

```diff
@@ -0,0 +1,290 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+import functools as _functools
+
+from traits.observers._observer_graph import (
+    ObserverGraph as _ObserverGraph,
+)
+
+
+class Expression:
+    """
+    Expression is an object for describing what traits are being observed
+    for change notifications. It can be passed directly to
+    ``HasTraits.observe`` method or the ``observe`` decorator.
+    """
+    def __init__(self):
+        # ``_levels`` is a list of list of IObserver.
+        # Each item corresponds to a layer of branches in the ObserverGraph.
+        # The last item is the most nested level.
+        # e.g. _levels = [[observer1, observer2], [observer3, observer4]]
+        # observer3 and observer4 are both leaf nodes of a tree, and they are
+        # "siblings" of each other. Each of observer3 and observer4 has two
+        # parents: observer1 and observer2.
+        # When ObserverGraph(s) are constructured from this expression, one
+        # starts from the end of this list, to the top, and then continues to
+        # the prior_expressions
+        self._levels = []
+
+        # Represent prior expressions to be combined in series (JOIN)
+        # or in parallel (OR). This is either an instance of _SeriesExpression
+        # or an instance of _ParallelExpression.
+        self._prior_expression = None
+
+    def __eq__(self, other):
+        """ Return true if the other value is an Expression with equivalent
+        content.
+
+        e.g. ``(trait("a") | trait("b")) == (trait("b") | trait("a"))`` will
```

Comment:
Do we need this extra complication? I think I'd expect equality for an `Expression` simply to be structural equality.

Reply:
This is merely for the users. As a user, I would expect `trait("a") | trait("b")` and `trait("b") | trait("a")` to be equivalent. It also does not hurt to make them not equal though. Nothing else depend on this.

Reply:
Overlooked this :(. See #1080